### PR TITLE
[#161850]Add ability to hide Price Groups

### PR DESCRIPTION
**Pull request by diebas**
# Release Notes

Added the ability to hide price groups in the admin menu. This option is only available for non-global price groups.
Once a price group is hidden, current price policies related to that group aren't visible or editable anymore. Old price policies remain visible. It also updates their can_parchase value to false

# Screenshot

Screenshot 2024-10-25 at 10 27 08 AM

Screenshot 2024-10-25 at 10 34 55 AM

# Additional Context

This is a draft PR, tests are on their way

### CODE REVIEW

File: config/locales/en.yml [hunk 1]
```diff
@@ -954,6 +954,7 @@ en:
   price_groups:
     price_group_fields:
       internal_instruction: "Check this box if this price group is internal:"
+      hidden_instruction: "Check this box if you want to hide this price group:"
```
Comment (LeticiaErrandonea) [inline config/locales/en.yml:957]: Maybe we should use another voice? Something like

```suggestion
      hidden_instruction: "Check this box if this price group is no longer used:"
```

File: db/migrate/20241014161738_add_is_hidden_to_price_groups.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class AddIsHiddenToPriceGroups < ActiveRecord::Migration[7.0]
+  def up
+    add_column :price_groups, :is_hidden, :boolean, default: false
+
+    PriceGroup.reset_column_information
+    PriceGroup.update_all(is_hidden: false)
```
[the remark this answers is not in the document]
Comment (LeticiaErrandonea) [reply db/migrate/20241014161738_add_is_hidden_to_price_groups.rb:6]: Agreed. We usually go with an empty class declared in the file. You can find examples in other migrations.

File: app/models/price_group.rb [hunk 3]
```diff
@@ -117,6 +118,15 @@ def setup_skip_review_price_policies
 
   private
 
+  def is_hidden_changed?
+    self.will_save_change_to_attribute?(:is_hidden)
```
Comment (LeticiaErrandonea) [inline app/models/price_group.rb:122]: nit: `self` is unnecessary

Also, do we need to wrap `will_save_change_to_attribute?`?

File: app/models/price_group.rb [hunk 4]
```diff
@@ -117,6 +118,15 @@ def setup_skip_review_price_policies
...
+  end
+
+  def update_hidden_price_group_policies
+    policies = self.price_policies.current
```
Comment (LeticiaErrandonea) [inline app/models/price_group.rb:126]: nit: `self` is unnecessary

File: app/services/price_policy_builder.rb [hunk 5]
```diff
@@ -94,7 +94,7 @@ def editable?
 
   def get_price_policies
     return [] unless editable?
-    facility.price_groups.map do |price_group|
+    facility.price_groups.where(is_hidden: false).map do |price_group|
```
Comment (LeticiaErrandonea) [inline app/services/price_policy_builder.rb:97]: Not a strong opinion but I think it'd be clearer to use a scope

**Review by LeticiaErrandonea — COMMENTED** [review]
Looking good!

Left a few comments

File: app/models/price_group.rb [hunk 6]
```diff
@@ -21,10 +21,12 @@ class PriceGroup < ApplicationRecord
   before_destroy { throw :abort if global? }
   before_destroy { price_policies.destroy_all } # Cannot be a dependent: :destroy because of ordering of callbacks
   before_create  ->(o) { o.display_order = 999 unless o.facility_id.nil? }
+  before_update  :update_hidden_price_group_policies, if: :will_save_change_to_is_hidden?
```
Comment (LeticiaErrandonea) [inline app/models/price_group.rb:24]: What if I'm changing from hidden to visible? As the code is now, we'll make a call to set `can_purchase` to `false`, but that's unnecessary. I think we should avoid getting to that part of the code

File: db/migrate/20241014161738_add_is_hidden_to_price_groups.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+class AddIsHiddenToPriceGroups < ActiveRecord::Migration[7.0]
```
Comment (LeticiaErrandonea) [inline db/migrate/20241014161738_add_is_hidden_to_price_groups.rb:1]: Please add `# frozen_string_literal: true`

File: spec/support/shared_examples/hidden_price_groups_shared_examples.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    expect(page).to have_content(base_price_group.name)
+    expect(page).not_to have_content(price_group_to_hide.name)
+
+    click_link "Add Pricing Rules"
+
+    expect(page).to have_content(base_price_group.name)
+    expect(page).not_to have_content(price_group_to_hide.name)
+  end
+end
```
Comment (LeticiaErrandonea) [inline spec/support/shared_examples/hidden_price_groups_shared_examples.rb:56]: nit: add newline at the end of the file

File: spec/support/shared_examples/hidden_price_groups_shared_examples.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  it "hides price policies related to that price group" do
+    visit send("facility_#{item_type}s_path", facility, item)
+    click_link item.name
+    click_link "Pricing"
+    click_link "Add Pricing Rules"
+
+    expect(page).to have_content(price_group_to_hide.name)
+
+    if item_type == "instrument"
```
Comment (LeticiaErrandonea) [inline spec/support/shared_examples/hidden_price_groups_shared_examples.rb:12]: nit: I wonder whether `item.is_a?(Instrument)` is easier to read

File: spec/system/admin/instrument_price_policies_controller_spec.rb [hunk 10]
```diff
@@ -154,6 +154,12 @@
         expect(page).to have_no_link "Remove"
       end
     end
+
+    describe "with hidden price policies", :js do
```
Comment (LeticiaErrandonea) [inline spec/system/admin/instrument_price_policies_controller_spec.rb:158]: Do we need `:js` here?

File: spec/system/admin/instrument_price_policies_controller_spec.rb [hunk 11]
```diff
@@ -399,5 +405,11 @@
       # External
       expect(page).not_to have_content("$90.00") # Step 3 rate (removed)
     end
+
+    describe "with hidden price policies", :js do
```
Comment (LeticiaErrandonea) [inline spec/system/admin/instrument_price_policies_controller_spec.rb:409]: Same as above regarding `:js`

File: spec/system/admin/item_price_policies_controller_spec.rb [hunk 12]
```diff
@@ -68,6 +68,10 @@
     end
   end
 
+  describe "with hidden price policies", :js do
```
Comment (LeticiaErrandonea) [inline spec/system/admin/item_price_policies_controller_spec.rb:71]: Same as above regarding `:js`

**Review by LeticiaErrandonea — COMMENTED** [review]
I really like the shared examples idea!

Left some comments

File: spec/system/admin/instrument_price_policies_controller_spec.rb (hunk 10 above)
Comment (diebas) [reply spec/system/admin/instrument_price_policies_controller_spec.rb:158]: Well, I can't tell the exact reason behind this, but if I remove the :js flag the tests fail. Capybara isn't able to find some fields

**Review by LeticiaErrandonea — APPROVED** [review]
LGTM!

Thanks for the changes!

**Review by giladshanan — APPROVED** [review]

**Pull request merged**